app2/ota_client.py

```diff
@@ -370,117 +375,113 @@ def _create_symbolic_links(self, list_file, standby_path):
-        lines = open(list_file).read().splitlines()
-        reginf_list = [RegularInf(l) for l in lines]
-        self._total_regular_files = len(reginf_list)  # via setter
+    def _create_regular_files(self, url_base: str, cookies, list_file, standby_path):
+
         with Manager() as manager:
-            hardlink_dict = manager.dict()
             error_queue = manager.Queue()
             # NOTE: manager.Value doesn't work properly.
             processed_list = manager.list()
 
             def error_callback(e):
                 error_queue.put(e)
 
+            boot_standby_path = self._ota_status.get_standby_boot_partition_path()
+            # bind the required options before we use this method
+            _create_regfile_func = partial(
+                self._create_regular_file,
+                url_base=url_base,
+                cookies=cookies,
+                standby_path=standby_path,
+                processed_list=processed_list,
+                boot_standby_path=boot_standby_path,
+            )
+
             with Pool() as pool:
-                self._create_regular_files_with_async(
-                    url_base,
-                    cookies,
-                    standby_path,
-                    reginf_list,
-                    hardlink_dict,
-                    processed_list,
-                    pool.apply_async,
-                    error_callback,
-                )
+                total_regular_files = 0
+                hardlink_dict = dict()  # sha256hash[tuple[reginf, event]]
+
+                # imap_unordered return a lazy itorator without blocking
+                # so we cannot use len() over it
+                reginf_list = pool.imap_unordered(RegularInf, open(list_file).read().splitlines())
+                for reginf in reginf_list:
+                    total_regular_files += 1
+
+                    if reginf.nlink >= 2:
+                        prev_reginf, event = hardlink_dict.setdefault(
+                            reginf.sha256hash, (reginf, manager.Event())
+                        )
+
+                        # multiprocessing.apply_async
+                        # input args:
+                        #   func, args: list, kwargs: dict, *, callback, error_callback
+                        # output:
+                        #   async_result
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf, prev_reginf),
+                            {"hardlink_event": event},
+                            error_callback=error_callback,
+                        )
+                    else:
+                        pool.apply_async(
+                            _create_regfile_func,
+                            (reginf,),
+                            error_callback=error_callback,
+                        )
+                self._total_regular_files = total_regular_files
```

Comment:
`self._total_regular_files is total` should be total number of regular files.

Reply:
Yes, I use this counter variable `total_regular_files ` to count the number of regular files in the for loop.

Reply:
> `self._total_regular_files is total` should be total number of regular files. Not number of regular files processed.

Yes I am doing what you said,
```
for reginf in reginf_list:
    total_regular_files += 1
```
is the same as `len(reginf_list)` in previous implementation.

Reply:
The reason why I count it in the for loop is as follow:
```
                # imap_unordered return a lazy itorator without blocking
                # so we cannot use len() over it
```

Reply:
:thinking: 
previous implementation:
```
self._total_regular_files = len(reginf_list) # via setter
````
I think you're saying `self._regular_files_processed`?